server/athenian/api/controllers/miners/github/pull_request.py

```diff
@@ -0,0 +1,207 @@
+        try:
+            return cls(func(arg.best for arg in args if arg.best is not None), None)
+        except ValueError:
+            return cls(None, None)
+
+
+@dataclass(frozen=True)
+class PullRequestTimes:
+    """Various PR update timestamps."""
+
+    created: Fallback[datetime]                                # PR_C
+    first_commit: Fallback[datetime]                           # PR_CC
+
+    @property
+    def work_begins(self) -> Fallback[datetime]:               # PR_B   noqa: D102
+        return Fallback.min(self.created, self.first_commit)
+
+    last_commit_before_first_review: Fallback[datetime]        # PR_CFR
+    last_commit: Fallback[datetime]                            # PR_L
+    merged: Fallback[datetime]                                 # PR_M
+    closed: Fallback[datetime]                                 # PR_CL
+    first_comment_on_first_review: Fallback[datetime]          # PR_W
+    first_review_request: Fallback[datetime]                   # PR_S
+    approved: Fallback[datetime]                               # PR_A
+    first_passed_checks: Fallback[datetime]                    # PR_VS
+    last_passed_checks: Fallback[datetime]                     # PR_VF
+    finalized: Fallback[datetime]                              # PR_F
+    released: Fallback[datetime]                               # PR_R
+
+
+class ReviewResolution(Enum):
+    """Possible review "state"-s in the metadata DB."""
+
+    APPROVED = "APPROVED"
+    CHANGES_REQUESTED = "CHANGES_REQUESTED"
+    COMMENTED = "COMMENTED"
+
+
+class PullRequestTimesMiner(PullRequestMiner):
+    """Extract the pull request update timestamps from the metadata DB."""
+
+    @classmethod
+    def _compile(cls, pr: pd.Series, reviews: pd.DataFrame, review_comments: pd.DataFrame):
+        created_at = Fallback(pr["created_at"], None)
+        merged_at = Fallback(pr["merged_at"], None)
+        first_comment_on_first_review = Fallback(review_comments["created_at"].min(), merged_at)
+        last_commit_before_first_review = first_comment_on_first_review  # FIXME(vmarkovtsev): no commit timestamps # noqa
+        first_review_request = Fallback(None, Fallback.min(
+            Fallback.max(created_at, last_commit_before_first_review),
+            first_comment_on_first_review))  # FIXME(vmarkovtsev): no review request info
+        if merged_at.value is not None:
+            reviews_before_merge = reviews[reviews["submitted_at"] <= merged_at.best]
+            grouped_reviews = reviews_before_merge \
+                .sort_values(["submitted_at"], ascending=True) \
+                .groupby("user_login", sort=False) \
+                .nth(0)  # the most recent review for each reviewer
+            if (grouped_reviews["state"] == ReviewResolution.CHANGES_REQUESTED).any():
+                # merged with negative reviews
+                approved_at_value = None
+            else:
+                approved_at_value = grouped_reviews[
```

Comment:
This can be `np.NaN` instead of `None` if there's no `ReviewResolution.APPROVED`.

Reply:
I've not checked how NaNs are serialized in JSON. Interesting, I'll change if they work fine.

Reply:
Due to the NaN check in `Fallback`, setting this to NaN will be equivalent to None.